rcl/test/rcl/allocator_testing_utils.h

```diff
@@ -87,6 +87,80 @@ set_failing_allocator_is_failing(rcutils_allocator_t & failing_allocator, bool s
   ((__failing_allocator_state *)failing_allocator.state)->is_failing = state;
 }
 
+typedef struct __time_bomb_allocator_state
+{
+  int count_until_failure;
+} __time_bomb_allocator_state;
+
+void * time_bomb_malloc(size_t size, void * state)
+{
+  __time_bomb_allocator_state * time_bomb_state = (__time_bomb_allocator_state *)state;
+  if (time_bomb_state->count_until_failure >= 0 &&
+    time_bomb_state->count_until_failure-- == 0)
+  {
+    return nullptr;
+  }
+  return rcutils_get_default_allocator().allocate(size, rcutils_get_default_allocator().state);
+}
+
+void *
+time_bomb_realloc(void * pointer, size_t size, void * state)
+{
+  __time_bomb_allocator_state * time_bomb_state = (__time_bomb_allocator_state *)state;
+  if (time_bomb_state->count_until_failure >= 0 &&
+    time_bomb_state->count_until_failure-- == 0)
+  {
+    return nullptr;
+  }
+  return rcutils_get_default_allocator().reallocate(
+    pointer, size, rcutils_get_default_allocator().state);
+}
+
+void
+time_bomb_free(void * pointer, void * state)
+{
+  __time_bomb_allocator_state * time_bomb_state = (__time_bomb_allocator_state *)state;
+  if (time_bomb_state->count_until_failure >= 0 &&
+    time_bomb_state->count_until_failure-- == 0)
+  {
+    return;
+  }
+  rcutils_get_default_allocator().deallocate(pointer, rcutils_get_default_allocator().state);
+}
+
+void *
+time_bomb_calloc(size_t number_of_elements, size_t size_of_element, void * state)
+{
+  __time_bomb_allocator_state * time_bomb_state = (__time_bomb_allocator_state *)state;
+  if (time_bomb_state->count_until_failure >= 0 &&
+    time_bomb_state->count_until_failure-- == 0)
+  {
+    return nullptr;
+  }
+  return rcutils_get_default_allocator().zero_allocate(
+    number_of_elements, size_of_element, rcutils_get_default_allocator().state);
+}
+
+static inline rcutils_allocator_t
+get_time_bombed_allocator(void)
+{
+  static __time_bomb_allocator_state state;
+  state.count_until_failure = 1;
```

Comment:
why 1?
I understand it works now, but it's heavily relying on an implementation detail.

Tests that rely on an implementation detail are harmful because when you refactor the code they either stop being useful and continue passing without checking anything useful, or because they start failing and you don't understand even what's the intention of the test.

Reply:
I didn't think much about the init value, the expected usage for this allocator involves making a call like `set_time_bombed_allocator_count(&alloc, count);`. Do you suggest me changing the default value to be 0 instead? (and fail in the first call).

Reply:
AFAIU, If you put zero it won't fail where you want, but I would fail here https://github.com/ros2/rcl/blob/632f0ac8c75888f82521587314fabaec3d940c9e/rcl/src/rcl/arguments.c#L86

Reply:
With this PR it fails for the three `BAD_ALLOC` cases, please see:
https://github.com/ros2/rcl/blob/632f0ac8c75888f82521587314fabaec3d940c9e/rcl/test/rcl/test_arguments.cpp#L1054-L1066

And you can see I called the `set_time_bombed_allocator_count` each time.

Reply:
ok, I think it's a bit fragile, but not too bad.